cores/esp8266/WString.h

```diff
@@ -275,41 +274,42 @@ class String {
         // parsing/conversion
         long toInt(void) const;
         float toFloat(void) const;
-	double toDouble(void) const;
+        double toDouble(void) const;
 
     protected:
         // Contains the string info when we're not in SSO mode
-        struct _ptr { 
+        struct _ptr {
             char *   buff;
             uint16_t cap;
             uint16_t len;
         };
         // This allows strings up up to 11 (10 + \0 termination) without any extra space.
-        enum { SSOSIZE = sizeof(struct _ptr) + 4 - 1 }; // Characters to allocate space for SSO, must be 12 or more
+        static constexpr auto SSOSIZE = sizeof(struct _ptr) + 4 - 1; // Characters to allocate space for SSO, must be 12 or more
         struct _sso {
             char     buff[SSOSIZE];
-            unsigned char len   : 7; // Ensure only one byte is allocated by GCC for the bitfields
-            unsigned char isSSO : 1;
+            unsigned char len      : 7; // Ensure only one byte is allocated by GCC for the bitfields
+            unsigned char isNotSSO : 1;
+            _sso() : len(0), isNotSSO(0) {
```

Comment:
direct member init for bitfield requires c++20, will bring some of CI tests failure (possibly host emulation)

Reply:
That's not what I meant. You should be able to do direct member init below in the union in some form or other.

Reply:
For bitfield members that's not possible, as that's one of the new features of C++20.
So for this special case of the bitfields you can only initialize them in the constructor of the `_sso` struct.

Reply:
As a quick test:
```cpp
  union
  {
    _sso sso = {{0},{0}};
    _ptr ptr;
  };
```

works fine with C++11.

Reply:
thx @devyte:

that's surely fine, in terms of the logic: wholly zeroed, due to funccall of `memset(this, 0, sizeof(_sso))`.
zeroing `_sso.buff[0]` and `_sso.`bitfields would be sufficient (and of course most efficient).

both class member initializers (`char member = 0;`) and braces initialization (`struct _foo foo = { 0, ... };`) cause whole initialization
(assumed that each of unspecified members has default value).

member initializer (`: bar(0), ...`) seems to be only way to realize sparse(holed) initialization.

Reply:
The sso struct is 12 bytes in size, yet you seem keen on doing microoptimizations here. I don't consider this worthwhile, the only case where this would be significant is in a tight loop where >90% of the time is spent in initing a String object, and in such a case I would likely yell at the author. Personally, I think there are other places where spending time and effort would bring far better returns. 
If you decide to pursue this, please:
- find bin size for initialization with {{0},{0}}
- compare against bin size with your member initializer
- compare runtime performance of both cases
- report back here

BTW, we're already on gcc 10.x. Although we have C++17 enabled, in theory we have access to C++20 experimental features.

Reply:
> BTW, we're already on gcc 10.x. Although we have C++17 enabled, in theory we have access to C++20 experimental features.

It was already mentioned before, this class is also ported to ESP32.
Are they also using gcc 10.x?
To be honest, if the braces initialization also works, then I would say to stick with what's working and not forcing to only be able to compile with the newest C++ standard.

What I'm more worried about is how the initialization is done when in a union.
In other words, what will be guaranteed to be initialized with the isSSO set to 1 (or the inverted one set to 0)?
We can talk about semantics and how the code may be the most aestetic pleasing to the reader.
But in the end it is important this one flag is always default set to mean `isSSO` is active.
If that should be done via automatic zero'ed initialization (thus inverting the flag) or the suggested:
```c++
  union
  {
    _sso sso = {{0},{0}};
    _ptr ptr;
  };
```
That's all fine by me.
Most important is that this `isSSO` is in its 'enabled' state (which isn't with this suggested code by the way) as default member value.